### PR TITLE
Provide openConnections method a blocking handler for plugins

### PR DESCRIPTION
**Pull request by wburns**
* HotRod new RemoteCacheManger would block event loop thread

Fixes https://github.com/Hyperfoil/Hyperfoil/issues/333

### CODE REVIEW

File: core/src/main/java/io/hyperfoil/core/impl/LocalSimulationRunner.java [hunk 1]
```diff
@@ -45,7 +46,15 @@ public void run() {
 
       CountDownLatch latch = new CountDownLatch(1);
       init();
-      openConnections(result -> latch.countDown());
+      openConnections(callable -> {
+               try {
+                  callable.call();
```
Comment (wburns) [inline core/src/main/java/io/hyperfoil/core/impl/LocalSimulationRunner.java:51]: I _think_ this okay as there is no event loop around with a local simulation run, but @franz1981 can you confirm?

Comment (franz1981) [reply core/src/main/java/io/hyperfoil/core/impl/LocalSimulationRunner.java:51]: Will tell you tomorrow, but ATM I would say no

Comment (franz1981) [reply core/src/main/java/io/hyperfoil/core/impl/LocalSimulationRunner.java:51]: Let me check how I could configure block hound for our tests

**Comment by franz1981** [comment]
I'm checking what's going on and found that:
1. https://github.com/Hyperfoil/Hyperfoil/blob/77e538713b169fcad0d7773ec3d64b098f68e260/clustering/src/main/java/io/hyperfoil/clustering/AgentVerticle.java#L265 is assuming that `assert context.isEventLoopContext()` which doesn't mean is running from the even loop (according to the doc of Vertx), but given that it is issued as consumer of the Vertx `EventBus` it should be running from the event loop
2. when `openConnections` is issued, it will call into `HotRodRunData::openConnections`, than https://github.com/Hyperfoil/Hyperfoil/blob/3d5d7751df384977d5e51e1709f83790e861495f/hotrod/src/main/java/io/hyperfoil/hotrod/connection/HotRodRemoteCachePoolImpl.java#L49 which, on `RemoteCacheManager`'s `start` can hit this:
```java
  java.lang.Thread.State: RUNNABLE
	  at org.infinispan.client.hotrod.impl.transport.netty.ChannelFactory.pingServersIgnoreException(ChannelFactory.java:197)
	  at org.infinispan.client.hotrod.impl.transport.netty.ChannelFactory.start(ChannelFactory.java:150)
	  at org.infinispan.client.hotrod.RemoteCacheManager.actualStart(RemoteCacheManager.java:362)
	  at org.infinispan.client.hotrod.RemoteCacheManager.start(RemoteCacheManager.java:316)
	  at org.infinispan.client.hotrod.RemoteCacheManager.<init>(RemoteCacheManager.java:174)
	  at org.infinispan.client.hotrod.RemoteCacheManager.<init>(RemoteCacheManager.java:131)
	  at io.hyperfoil.hotrod.connection.HotRodRemoteCachePoolImpl.start(HotRodRemoteCachePoolImpl.java:49)
	  at io.hyperfoil.hotrod.HotRodRunData.openConnections(HotRodRunData.java:55)
	  at io.hyperfoil.core.impl.SimulationRunner.openConnections(SimulationRunner.java:163```
which await `PingOperation`s to reach out each server, while running it from the Netty event loop,see
```java
private void pingServersIgnoreException() {
      Collection<InetSocketAddress> servers = topologyInfo.getAllServers();
      for (SocketAddress addr : servers) {
         // Go through all statically configured nodes and force a
         // connection to be established and a ping message to be sent.
         try {
            await(fetchChannelAndInvoke(addr, operationsFactory.newPingOperation(true)));
         } catch (Exception e) {
            // Ping's objective is to retrieve a potentially newer
            // version of the Hot Rod cluster topology, so ignore
            // exceptions from nodes that might not be up any more.
            if (log.isTraceEnabled())
               log.tracef(e, "Ignoring exception pinging configured servers %s to establish a connection",
                     servers);
         }
      }
   }
   ```
which `PingOperation` is going to be executed as soon as the connection to servers are established, but running from the event loop thread configured:
```java
"epollEventLoopGroup-5-1@9906" tid=0x35 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at org.infinispan.client.hotrod.impl.operations.PingOperation.invoke(PingOperation.java:53)```
```
In short, my perception is that `openConnections` doesn't assume to be non-blocking, so I'm proposing a different fix here,ie https://github.com/franz1981/Hyperfoil/tree/never_blocking_el

IDK if is better or not, but I see that the local runnner doesn't assume opeening the connections to be a non-blocking operation, so I see as a responsability of the caller (which knows if is running on an event loop - which should never be blocked) to handle such.
Let me know if this is still fixing your case; I'm trying to write a test for this, regardless (both fixes should work with the test)

**Comment by wburns** [comment]
@franz1981 while I am certain your commit will fix the issue, I personally don't like the approach. The openConnections Plugin API already implies it should be non blocking with the method declaration as shown here: 
https://github.com/Hyperfoil/Hyperfoil/blob/ac3ad2eaa929d363d8f7e850f6c1d4d40b594ab4/core/src/main/java/io/hyperfoil/core/api/PluginRunData.java#L15

The argument to me says the implementor shouldn't block and pass any operations that could block to the consumer so it can be notified after they are all done.

**Comment by franz1981** [comment]
Yeah, but at this point, if we assume that it is the duty of the implementor to not block (because its API suggest that it shouldn't be blocking), you shouldn't allow to pass vertx::executeBlocking to it, because it would, again, give the responsibility to the caller, despite the async API.

If we want openConnections to be non blocking, it should be the spwcofoc HotRod implementation thst should take care to do it (being non blocking as expected) without changing its API.

In addition to what said, I would add some comment, on the API, the caller, or the implementor that explain what's going and why.

**Comment by wburns** [comment]
> Yeah, but at this point, if we assume that it is the duty of the implementor to not block (because its API suggest that it shouldn't be blocking - I have no strong opinion on it TBH, and probably only @rvansa knows it), you shouldn't allow to pass vertx::executeBlocking to it, because it would, again, give the responsibility to the caller, despite the expected async API (& behavior).

> If we want openConnections to be non blocking, it should be the HotRod implementation that should take care to do it (being non blocking as expected) without changing its API (or via some other base class which provides the mechanism to do it right).

At this point implementors have to be non blocking to work properly. I guess we could relax the requirements for lifecycle operations such as start and stop.

The problem about forcing the implementation to handle the blocking is, we do not provide the tools to them. Namely there is no access to a blocking executor or way to run blocking code. This in turn means the implementor has to create their own threads, which seems very unfriendly. The nicest way to me was to provide a facility to run some code in a blocking thread (aka. vertx::executeBlocking). I chose that abstraction to not expose the plugin to vertx, but I am open to other avenues if you want.

> Clearly if we assume openConnections to be async, we should (if broken) fix the close side as well, to make sure the state machine is right (not in this PR, unless making it non blocking mess up with its state).

Yeah, I had debated about changing close as well, as that would break the non blocking contract as well. However, it doesn't actively cause the test to not run properly, so I left it for another day.

> In addition to what said, I would add some comment, on the API, the caller, or the implementor that explain what's going on and why.

Definitely agree, I didn't have a chance to add the comment yesterday as I had to leave immediately after I submitted. I can add this to the PR though, for sure.


TLDR; I guess I am fine with making start/stop run on a blocking thread and allowing the implementation to be blocking in those cases, it just seems inconsistent as everything else has to be non blocking in the plugin.

**Comment by franz1981** [comment]
Yeah I tried to check if we pass somehow any blocking executor (from vertx) on the implementors but seems not :/ 
So I agree that we have to use a generic mechanism (or an executor-like parameter too, at this point is the same..);

I have noticed that:
- HotRodRunData::openConnections was ignoring the passed promiseCollector (meaning that if something break - now asynchronously, none knows...)
- HotRodRemoteCachePoolImpl::start is manipulating thread unsafe maps eg remoteCaches, which could be accessed by HotRodRequestStep::invoke at pool.getRemoteCache; if we are now asynx we risk thst the connection won't be established, the access will race AND if lucky (while unlucky) no remoteCache will be found likely throwing NPE on the step (which, remember, has started regardless the pool start completion!)

**Comment by wburns** [comment]
> Yeah I tried to check if we pass somehow any blocking executor (from vertx) on the implementors but seems not :/ So I agree that we have to use a generic mechanism (or an executor-like parameter too, at this point is the same..);
> 
> I have noticed that:
> 
> * HotRodRunData::openConnections was ignoring the passed promiseCollector (meaning that if something break - now asynchronously, none knows...)

It should throw an exception at least, but yes I agree and is fixed in this PR.

> * HotRodRemoteCachePoolImpl::start is manipulating thread unsafe maps eg remoteCaches, which could be accessed by HotRodRequestStep::invoke at pool.getRemoteCache; if we are now async, we risk that the connection won't be established, the access will race AND if lucky (while unlucky) no remoteCache will be found likely throwing NPE on the step (which, remember, has started regardless the pool start completion!)

I am not certain, but I don't think Invoke can be ran concurrently with start as openConnections doesn't call handle until all the pools are created. However, the variable is indeed not thread safe so there is a small chance that the other thread may not see the updated value, although I doubt it wouldn't hit a memory barrier to sync that up someplace else. But we can easily change the map to a ConcurrentMap to be safe.

**Comment by franz1981** [comment]
Yeah I tried to check if we pass somehow any blocking executor (from vertx) on the implementors but seems not :/ 
So I agree that we have to use a generic mechanism (or an executor-like parameter too, at this point is the same..);

I have noticed that:
- HotRodRunData::openConnections was ignoring the passed promiseCollector (meaning that if something break - now asynchronously, none knows...)
- HotRodRemoteCachePoolImpl::start is manipulating thread unsafe maps eg remoteCaches, which could be accessed by HotRodRequestStep::invoke at pool.getRemoteCache; if we are now asynx we risk thst the connection won't be established, the access will race AND if lucky (while unlucky) no remoteCache will be found likely throwing NPE on the step (which, remember, has started regardless the pool start completion!)

**Comment by franz1981** [comment]
I can easily verify if NPE can happen actually, let me use your branch

**Comment by rvansa** [comment]
>  if we assume that it is the duty of the implementor to not block (because its API suggest that it shouldn't be blocking - I have no strong opinion on it TBH, and probably only @rvansa knows it), you shouldn't pass any vertx::executeBlocking to it, because it would, again, give the responsibility to the caller, despite the expected async API (& behavior).

Yes, I assumed this is running from Vert.x event-loop and hence shouldn't block. I don't see any issue with passing a blocking executor as long as eventually the callback is invoked. As this is executed only during initialization, you won't break any benchmark by changing this - as long as you don't break the Vert.x wiring.

> The problem about forcing the implementation to handle the blocking is, we do not provide the tools to them. Namely there is no access to a blocking executor or way to run blocking code. This in turn means the implementor has to create their own threads, which seems very unfriendly.

Initially the API was written for HTTP, and async implementation in general, assuming that any reasonable network protocol API will be non-blocking. It's a defect on the target library side to let the thread wait :)

These callback-based APIs look a bit unwieldy anyway, though - change them at will if that suits you better :)

**Comment by franz1981** [comment]
Many thanks @rvansa for it!

**Comment by wburns** [comment]
Added the docs for the openConnections SPI method. Unfortunately, after tomorrow I will not be able to continue working on this.

**Comment by franz1981** [comment]
No need, you can merge it as it is

**Comment by franz1981** [comment]
Or just wait till I check what's left, if any (I cannot see what could be)

**Review by franz1981 — APPROVED** [review]
LGTM!

**Comment by franz1981** [comment]
Thanks for this fix, @wburns and well done, it was tricky to find!
